src/blog/tanstack-db-0.1-the-embedded-client-database-for-tanstack-query.md

```diff
@@ -0,0 +1,378 @@
+
+**TanStack DB enables Option C – Normalized Collections + Incremental Joins:** Load normalized collections through fewer API calls, then perform lightning-fast incremental joins in the client. You get the network efficiency of broad data loading with sub-millisecond query performance that makes Option A unnecessary.
+
+Instead of this:
+
+```typescript
+// View-specific API call every time you navigate
+const { data: projectTodos } = useQuery(
+  ['project-todos', projectId],
+  () => fetchProjectTodosWithUsers(projectId)
+)
+```
+
+You can do this:
+
+```typescript
+// Load normalized collections upfront (3 broader calls)
+const todoCollection = createQueryCollection({
+  queryKey: ['todos'],
+  queryFn: fetchAllTodos,
+})
+const userCollection = createQueryCollection({
+  queryKey: ['users'],
+  queryFn: fetchAllUsers,
+})
+const projectCollection = createQueryCollection({
+  queryKey: ['projects'],
+  queryFn: fetchAllProjects,
+})
+
+// Navigation is instant — no new API calls needed
+const { data: activeProjectTodos } = useLiveQuery(
+  (q) =>
+    q
+      .from({ t: todoCollection })
+      .innerJoin(
+        { u: userCollection },
+        ({ t, u }) => eq(t.userId, u.id)
+      )
+      .innerJoin(
+        { p: projectCollection },
+        ({ u, p }) => eq(u.projectId, p.id)
+      )
+      .where(({ t }) => eq(t.active, true))
+      .where(({ p }) =>
+        eq(p.id, currentProject.id)
+      )
+)
+```
+
+Now, clicking between projects, users, or views requires **zero API calls**. All the data is already loaded. New features like "show user workload across all projects" work instantly without touching your backend.
+
+Your API becomes simpler. Your network calls drop dramatically. Your frontend gets faster as your dataset grows.
+
+## The 20MB Question
+
+**Your app would be dramatically faster if you just loaded 20MB of normalized data upfront** instead of making hundreds of small API calls.
+
+Companies like Linear, Figma, and Slack load massive datasets into the client and achieve incredible performance through heavy investment in custom indexing, differential updates, and optimized rendering. These solutions are too complex and expensive for most teams to build.
+
+**TanStack DB brings this capability to everyone** through differential dataflow — a technique that only recomputes the parts of queries that actually changed. Instead of choosing between "many fast API calls with network waterfalls" or "few API calls with slow client processing," you get the best of both options: fewer network round-trips AND sub-millisecond client-side queries, even with large datasets.
+
+This isn't just about sync engines like Electric (though they make this pattern incredibly powerful). It's about enabling a fundamentally different data loading strategy that works with any backend — REST, GraphQL, or real-time sync.
```

Comment:
I would hyperlink Electric, as a courtesy to readers who don't know what it is.

Reply:
Actually perhaps better to link to the Electric TanStack DB blog post here -- maybe `[incredibly powerful](https://electric-sql.com/blog/2025/07/24/local-first-sync-with-tanstack-db)`?